readers/mutation_reader.cc

```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
-    return (*this)(to_mutation_fragment_kind_v2(kind), pos);
-}
 
-bool mutation_fragment_stream_validator::operator()(const mutation_fragment_v2& mf) {
-    const auto valid = (*this)(mf.mutation_fragment_kind(), mf.position());
-    if (valid && mf.is_range_tombstone_change()) {
-        _current_tombstone = mf.as_range_tombstone_change().tombstone();
-    }
-    return valid;
-}
-bool mutation_fragment_stream_validator::operator()(const mutation_fragment& mf) {
-    return (*this)(to_mutation_fragment_kind_v2(mf.mutation_fragment_kind()), mf.position());
-}
+    auto valid = true;
 
-bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind kind) {
-    bool valid = true;
+    // Check fragment kind order
     switch (_prev_kind) {
         case mutation_fragment_v2::kind::partition_start:
-            valid = kind != mutation_fragment_v2::kind::partition_start;
+            valid &= kind != mutation_fragment_v2::kind::partition_start;
             break;
         case mutation_fragment_v2::kind::static_row: // fall-through
         case mutation_fragment_v2::kind::clustering_row: // fall-through
         case mutation_fragment_v2::kind::range_tombstone_change:
-            valid = kind != mutation_fragment_v2::kind::partition_start &&
+            valid &= kind != mutation_fragment_v2::kind::partition_start &&
                     kind != mutation_fragment_v2::kind::static_row;
             break;
         case mutation_fragment_v2::kind::partition_end:
-            valid = kind == mutation_fragment_v2::kind::partition_start;
+            valid &= kind == mutation_fragment_v2::kind::partition_start;
             break;
     }
-    if (valid) {
-        _prev_kind = kind;
+    if (!valid) {
+        return false;
+    }
+
+    if (pos && _prev_kind != mutation_fragment_v2::kind::partition_end) {
+        auto cmp = position_in_partition::tri_compare(_schema);
+        auto res = cmp(_prev_pos, *pos);
+        if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
+            valid &= res <= 0;
+        } else {
+            valid &= res < 0;
+        }
+        if (!valid) {
+            return false;
+        }
+    }
+
+    _prev_kind = kind;
+    if (pos) {
+        _prev_pos = *pos;
+    } else {
+        switch (kind) {
+            case mutation_fragment_v2::kind::partition_start:
+                _prev_pos = position_in_partition(position_in_partition::partition_start_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::static_row:
+                _prev_pos = position_in_partition(position_in_partition::static_row_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::clustering_row:
+            case mutation_fragment_v2::kind::range_tombstone_change: // fall-through
+                if (!_prev_pos.has_clustering_key()) { // don't move pos if it is already a clustering one
```

Comment:
so how the current `pos` ever gets into prev_pos?

Reply:
See the true branch above: if `pos` is engaged it always gets into `_prev_pos`.

Reply:
I presume this would handle dummy rows.

Reply:
Actually it is better to check for partition-region instead here. Not all clustering positions have a key.

Reply:
agreed